contracts/core/Allo.sol

```diff
@@ -1,79 +1,359 @@
+        bool _cloneAllocationStrategy,
+        bool _cloneDistributionStrategy
+    ) external payable returns (uint256 poolId) {
+        /// Check the boolean flag to see if the strategy should be cloned
+        /// Additionally, force clone the strategy if it's approved
+        address allocationStrategy =
+            _cloneAllocationStrategy ? Clone.createClone(_allocationStrategy, _nonce++) : _allocationStrategy;
+
+        address distributionStrategy =
+            _cloneDistributionStrategy ? Clone.createClone(_distributionStrategy, _nonce++) : _distributionStrategy;
+
+        return _createPool(_identityId, allocationStrategy, payable(distributionStrategy), _token, _amount, _metadata);
+    }
 
-    // Public getter on the pools mapping
-    /// @notice calls out to the registry to get the identity metadata
-    function getPoolInfo(uint256 _poolId) external view returns (PoolData memory, string memory) {
-        // Implement the function here
+    /// @notice Creates a new pool
+    /// @param _identityId The identityId of the pool creator in the registry
+    /// @param _allocationStrategy The address of the allocation strategy
+    /// @param _distributionStrategy The address of the distribution strategy
+    /// @param _token The address of the token that the pool is denominated in
+    /// @param _amount The amount of the token to be deposited into the pool
+    /// @param _metadata The metadata of the pool
+    function createPool(
+        bytes32 _identityId,
+        address _allocationStrategy,
+        address payable _distributionStrategy,
+        address _token,
+        uint256 _amount,
+        Metadata memory _metadata
+    ) external payable returns (uint256 poolId) {
+        return _createPool(_identityId, _allocationStrategy, _distributionStrategy, _token, _amount, _metadata);
     }
 
-    // @todo insert clonable strategy library, including validation that an existing pool has safe strategy
+    /// @notice Creates a new pool
+    /// @param _identityId The identityId of the pool creator in the registry
+    /// @param _allocationStrategy The address of the allocation strategy
+    /// @param _distributionStrategy The address of the distribution strategy
+    /// @param _token The address of the token that the pool is denominated in
+    /// @param _amount The amount of the token to be deposited into the pool
+    /// @param _metadata The metadata of the pool
+    function _createPool(
+        bytes32 _identityId,
+        address _allocationStrategy,
+        address payable _distributionStrategy,
+        address _token,
+        uint256 _amount,
+        Metadata memory _metadata
+    ) internal returns (uint256 poolId) {
+        if (!registry.isOwnerOrMemberOfIdentity(_identityId, msg.sender)) {
+            revert NO_ACCESS_TO_ROLE();
+        }
+
+        if (usedStrategies[_allocationStrategy] || usedStrategies[_distributionStrategy]) {
+            revert STRATEGY_ALREADY_USED();
+        }
```

Comment:
Hmmm the risk here is that someone could frontrun your transaction and use your strategies to DOS you. Is there some data that will always be stored on the strategy that could be checked here to make sure it's the right user calling it?

Once we're needing to do that, we might just want to make strategies verify that they are unused themselves. Like, for example, enforce an `initialize()` function as a part of this call with proper validation to make sure it has to be used by the right person (ie `require(msg.sender == Allo.sol && caller == owner());` or something)

Sorry for back and forth on this. Hadn't thought about the DOS risk until I saw it.

Any ideas on how to do this check elegantly in a way that strategy deployers won't screw up?

Reply:
> Once we're needing to do that, we might just want to make strategies verify that they are unused themselves. Like, for example, enforce an initialize() function as a part of this call with proper validation to make sure it has to be used by the right person (ie require(msg.sender == Allo.sol && caller == owner()); or something)

Is this the same as :
- Interfaces now become an abstract contract
- Having modifiers to offer those checks   `require(msg.sender == Allo.sol && caller == owner());`

Reply:
Did we decide that we will enforce all pool creators having identities? I think yes, right?

If that's the case, this solution seems right to me.

Reply:
Reopening this conversation. I think @KurtMerbeth has the right solution, but doesn't look like it's been implemented.

We would need a check here that `_identityId == strategy.identityId()` for both strategies.

While we're doing that, it might make more sense to:
- Get rid of the `usedStrategies` mapping
- Call the `initialize()` function on the strategies straight from the pool
- In that function, check that `initialized = false` (which accomplishes same as `usedStrategies`
- Any non clone strategy will, in that function, check that `_identityId (argument) == identityId (locally stored)` (same as Kurt's recommendation, where `identityId` should be set in constructor
- Sets `intialized = true` and whatever else is needed

This means we should probably also add a generic `bytes memory` field when creating a pool that gets passed as initialize data.